src/recode.jl

```diff
@@ -52,22 +54,47 @@ A user defined type could override this method to define an appropriate test fun
 optimize_pair(pair::Pair) = pair
 optimize_pair(pair::Pair{<:AbstractArray}) = Set(pair.first) => pair.second
 
+function missing_check(value)
+    ismissing(value) && throw(MissingException("missing value found, but dest does not support them: " *
+                                       "recode them to a supported value"))
+    value
+end
+
+function recode!(dest::AbstractArray{T}, src::CategoricalArray, default::Any, pairs::Pair...) where {T}
+    if length(dest) != length(src)
+        throw(DimensionMismatch("dest and src must be of the same length (got $(length(dest)) and $(length(src)))"))
+    end
+    
+    pairs = map(pairs) do p
+        p.first => convert(T, p.second)
+    end
+    recoded = recode(src, default, pairs...)
+    if T >: Missing
+        dest .= unwrap.(recoded)
+    else
+        dest .= missing_check.(unwrap.(recoded))
+    end
+    dest
+end
+
 function recode!(dest::AbstractArray{T}, src::AbstractArray, default::Any, pairs::Pair...) where {T}
     if length(dest) != length(src)
         throw(DimensionMismatch("dest and src must be of the same length (got $(length(dest)) and $(length(src)))"))
     end
 
     opt_pairs = map(optimize_pair, pairs)
 
-    @inbounds for i in eachindex(dest, src)
-        x = src[i]
+    map!(dest, src) do x
 
-        for j in 1:length(opt_pairs)
-            p = opt_pairs[j]
-            # we use isequal and recode_in because we cannot really distinguish scalars from collections
-            if x ≅ p.first || recode_in(x, p.first)
-                dest[i] = p.second
-                @goto nextitem
+        # we use isequal and recode_in because we cannot really distinguish scalars from collections
+        for p in opt_pairs
+            if x ≅ p.first
+                return p.second
+            end
+        end
+        for p in opt_pairs
+            if recode_in(x, p.first)
+                return p.second
             end
```

Comment:
Since this was a while a go I'll need to take some time with it and rerun my (micro-)benchmarks to be sure but unless my memory fails me `recode_in` was a performance bottleneck and splitting the checks (in addition to switching to `map!`) made a noticeable difference for highly optimizable cases. You're absolutely right that it is a breaking change, and should have been highlighted in the PR since it warrants discussion. I spent some time trying to get `recode_in` to optimize away but was not satisfied with the result. The most troublesome part is of course the `any(x ≅ y for y in collection)` for the case when `collection` is a primitive. I'll get back to you with data.

Reply:
OK. Maybe better do this in a separate PR since it's a bit more tricky.